internal/components/component.go

```diff
@@ -34,37 +35,44 @@ var (
 	PortNotFoundErr       = errors.New("port should not be empty")
 )
 
-// PortParser is a function that returns a list of servicePorts given a config of type T.
-type PortParser[T any] func(logger logr.Logger, name string, o *Option, config T) ([]corev1.ServicePort, error)
-
 type PortRetriever interface {
 	GetPortNum() (int32, error)
 	GetPortNumOrDefault(logr.Logger, int32) int32
 }
 
-type Option struct {
+// PortParser is a function that returns a list of servicePorts given a config of type T.
+type PortParser[T any] func(logger logr.Logger, name string, defaultPort *corev1.ServicePort, config T) ([]corev1.ServicePort, error)
+
+// RBACRuleGenerator is a function that generates a list of RBAC Rules given a configuration of type T
+// It's expected that type T is the configuration used by a parser.
+type RBACRuleGenerator[T any] func(logger logr.Logger, config T) ([]rbacv1.PolicyRule, error)
+type PortBuilderOption[T any] func(*Option[T])
+
+type Option[T any] struct {
 	protocol    corev1.Protocol
 	appProtocol *string
 	targetPort  intstr.IntOrString
 	nodePort    int32
 	name        string
 	port        int32
+	portParser  PortParser[T]
+	rbacGen     RBACRuleGenerator[T]
 }
```

Comment:
Note: I need the generics for the parsing stuff, but it really crowds the type signatures. (see the receivers/helpers.go file) I could remove the functions from the options struct, and put it in the constructor, but that feels counter intuitive to the options struct.